sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala

```diff
@@ -318,25 +292,32 @@ private[sql] case class DataSourceScan(
         |   }
         | }""".stripMargin)
 
-      val value = ctx.freshName("value")
       s"""
          | if ($batch != null) {
          |   $scanBatches();
-         | } else if ($input.hasNext()) {
-         |   Object $value = $input.next();
-         |   if ($value instanceof $columnarBatchClz) {
-         |     $batch = ($columnarBatchClz)$value;
+         | } else {
+         |   long $getBatchStart = System.nanoTime();
+         |   if ($input.hasNext()) {
+         |     $batch = ($columnarBatchClz)$input.next();
+         |     $scanTimeTotalNs += System.nanoTime() - $getBatchStart;
          |     $scanBatches();
-         |   } else {
-         |     $scanRows((InternalRow) $value);
          |   }
          | }
        """.stripMargin
     } else {
+      val row = ctx.freshName("row")
+      val exprRows =
+        output.zipWithIndex.map(x => new BoundReference(x._2, x._1.dataType, x._1.nullable))
+      ctx.INPUT_ROW = row
+      ctx.currentVars = null
+      val columnsRowInput = exprRows.map(_.gen(ctx))
+      val inputRow = if (outputUnsafeRows) row else null
       s"""
-         |if ($input.hasNext()) {
-         |  $scanRows((InternalRow) $input.next());
-         |}
+         | while (!shouldStop() && $input.hasNext()) {
```

Comment:
I tried a few variants and didn't see a difference. I think only the batched scan is fast enough to be sensitive to this.